ClemBot.Bot/bot/cogs/search_cog.py

```diff
@@ -0,0 +1,159 @@
+import logging
+import re
+import json
+import aiohttp
+import discord
+import discord.ext.commands as commands
+import bot.extensions as ext
+from enum import Enum
+from markdownify import markdownify
+from urllib.parse import urlencode
+from urllib.parse import unquote_plus
+from bot.consts import Colors
+from bot.messaging.events import Events
+
+log = logging.getLogger(__name__)
+max_related_topics = 5
+image_url = 'https://duckduckgo.com'
+search_url = 'https://api.duckduckgo.com/'
+icon_url = 'https://i.imgur.com/Knq2kVa.png'
+related_topics_pattern = re.compile('(https://duckduckgo.com/)([acdne]/)?(\\S+)', re.IGNORECASE)
+
+
+class Category(Enum):
+    ARTICLE = 'A', 'Article'
+    DISAMBIGUATION = 'D', 'Disambiguation'
+    CATEGORY = 'C', 'Category'
+    NAME = 'N', 'Name'
+    EXCLUSIVE = 'E', 'Exclusive'
+    NONE = 'None', 'None'
+
+    def code(self):
+        return self.value[0]
+
+    def name(self):
+        return self.value[1]
+
+
+class SearchResult:
+
+    def __init__(self, json_response):
+        self.json_response = json_response
+
+    def __getitem__(self, item):
+        return self.json_response
+
+    def has_result(self):
+        return len(self.json_response['Heading']) > 0
+
+    def title(self):
+        return self.json_response['Heading']
+
+    def abstract(self):
+        return markdownify(self.json_response['Abstract'])
+
+    def category(self):
+        return category_from_code(self.json_response['Type'])
+
+    def has_thumbnail(self):
+        return len(self.json_response['Image']) > 0
+
+    def thumbnail(self):
+        # surprisingly 'Image' in the JSON returns the latter half of the URL,
+        # so we need to concat the other half, i.e., https://duckduckgo.com
+        return image_url + self.json_response['Image']
+
+    def related_topics(self) -> list:
+        # related topics can return topics that we don't actually want,
+        # so we're taking the subset of data that we can use and returning it
+        topics = list()
+        related_topics = self.json_response['RelatedTopics']
+        for i in range(0, len(related_topics)):
+            topic = related_topics[i]
+            if 'FirstURL' not in topic:
+                break
+            topics.append(topic)
+        return topics
+
+    def has_related_topics(self):
+        return len(self.related_topics()) > 0
+
+    def related_topics_formatted(self):
+        related_topics = self.related_topics()
+        related_topics_formatted = list[str]()
```

Comment:
This isnt valid python code

Reply:
I saw something somewhere that the []'s were for generics. I guess that was wrong lmao. Changed in latest commit.